File: src/algorithms/sorting/novelty/bogo-sort/bogo-sort.test.ts
import { describe, it, expect } from "vitest";
import { bogoSort } from "./sources/bogo-sort.ts?fn";

describe("bogoSort", () => {
  it("sorts a small array using seeded PRNG", () => {
    const result = bogoSort([3, 1, 2]);
    expect(result).toEqual([1, 2, 3]);
  });

  it("handles an already sorted array", () => {
    expect(bogoSort([1, 2, 3])).toEqual([1, 2, 3]);
  });

  it("handles a single element array", () => {
    expect(bogoSort([42])).toEqual([42]);
  });

  it("handles an empty array", () => {
    expect(bogoSort([])).toEqual([]);
  });

  it("does not mutate the original array", () => {
    const original = [3, 1, 2];
    const sorted = bogoSort(original);
    expect(sorted).toEqual([1, 2, 3]);
    expect(original).toEqual([3, 1, 2]);
  });

  it("produces a sorted result (within cap)", () => {
    const result = bogoSort([2, 1]);
    // Either sorted (if PRNG finds it within 100 iterations) or same array
    const isSorted = result[0]! <= result[1]!;
    expect(Array.isArray(result)).toBe(true);
    expect(result.length).toBe(2);
    // With seed 42 and only 2 elements, it should sort quickly
    expect(isSorted).toBe(true);
  });
});

File: src/algorithms/sorting/novelty/bogo-sort/sources/bogo_sort_test.rs
include!("bogo-sort.rs");

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn sorts_small_array_using_seeded_prng() {
        assert_eq!(bogo_sort(&[3, 1, 2]), vec![1, 2, 3]);
    }

    #[test]
    fn handles_already_sorted_array() {
        assert_eq!(bogo_sort(&[1, 2, 3]), vec![1, 2, 3]);
    }

    #[test]
    fn handles_single_element_array() {
        assert_eq!(bogo_sort(&[42]), vec![42]);
    }

    #[test]
    fn handles_empty_array() {
        assert_eq!(bogo_sort(&[]), vec![]);
    }

    #[test]
    fn does_not_mutate_original_array() {
        let original = vec![3, 1, 2];
        let sorted = bogo_sort(&original);
        assert_eq!(sorted, vec![1, 2, 3]);
        assert_eq!(original, vec![3, 1, 2]);
    }

    #[test]
    fn produces_sorted_result_within_cap() {
        let result = bogo_sort(&[2, 1]);
        assert_eq!(result.len(), 2);
        // With seed 42 and only 2 elements, it should sort quickly
        assert!(result[0] <= result[1]);
    }
}
